A new tally compute to obtain virial heat flow from group 2 to group 1

donatas-surblys opened the pull request:
**Summary**

A tally compute that obtains the heat flow into group 1 due to interaction with group 2. The "flux" naming is kept to keep consistency with already present computes.

For example, the following command calculates the virial heat flow (scalar) from the Wall into Liquid:

> compute hflow Liquid heat/flux/virial/tally Wall

In this example, as no particles cross the Wall-Liquid interface, the heat flow value will be identical to that obtained via method of planes method.  Dividing it by the cross-plane section will provide the heat flux.

I think this would be a good addition to LAMMPS, as currently there's no direct way to obtain heat flux over an interface.  If this compute is accepted I can update the tally documentation and also clarify what the original compute heat/flux/tally does, as currently the documentation is not clear enough and one has to read the source code to be sure. 


**Related Issue(s)**

PR #2501 would enable dynamic groups to be used, vastly increasing the utility of the compute.

**Author(s)**

Donatas Surblys (Tohoku University) surblys.donatas at gmail dot com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Full backward compatibility.

**Implementation Notes**

As a basic case, heat flow in the z direction over a control surface perpendicular to z, provided there are no particles crossing it, can be obtained as:

>  sum(Fij*(vi+vj)*zij/|zij|)

This compute generalizes the `zij/|zij|`, so that heat flow from group 2 to group 1 could be computed.  Basically, it is a slightly modified compute force/tally, that computes the `Fij*zij/|zij|` part during tallying, and multiplies by velocity and sums when the scalar value is requested. The `Fij*zij/|zij|` values are stored as per-atom 3 component arrays, similarly to stress/atom.  Summing those values will provide the total force on group 1 due to group 2, which would be identical to what the compute group/group provides.  In case of the earlier example with the Wall and Liquid, this can be used to obtain the force over the interface, with the sign depending on the relative orientation of the groups. Assuming the interface is perpendicular to z:

> compute iforce all reduce sum c_hflow[3]

Also, as I don't know a good way to handle it, the compute errors out if any particles are in both groups.  

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

akohlmey reviewed (changes requested):
Before this can be merged, it needs to be documented and the documentation integrated. If you also want to clarify the existing docs, that would be welcome, too. Since this is doing a complex task a (small) example demonstrating a typical use (and possibly contrasting to the other heat/flux/tally compute) would be appreciated as well.

donatas-surblys commented:
I've updated the compute tally documentation to include brief description of both *heat/flux/tally* and the new *heat/flux/virial/tally*. I didn't add too much detail, as the *heat/flux* compute already is quite extensive in this regard. I've also tried to demonstrate the difference between the three heat/flux* computes, and also demonstrated similarity between *group/group*, *force/tally* and *heat/flux/virial/tally* with a sample snippet, although it got slightly long.

I've also added the necessary links to the new compute, and added short descriptions to all of the tally computes in the compute.rst file.

sjplimp commented:
excellent context for the user - thanks

donatas-surblys commented:
Great! I just noticed that the example snippet in the updated documentation would be more consistent if the groups in heat/flux/virial/tally were swapped:
```
diff -r 3344d79ca368 -r 783c6ad74dc6 doc/src/compute_tally.rst
--- a/doc/src/compute_tally.rst	Tue Jul 20 18:03:11 2021 +0900
+++ b/doc/src/compute_tally.rst	Wed Jul 21 12:27:37 2021 +0900
@@ -106,7 +106,7 @@
    # Has LeftLiquid and RightWall groups along x direction.
 
    # Heat flux over the solid-liquid interface
-   compute hflow_hfvt LeftLiquid heat/flux/virial/tally RightWall
+   compute hflow_hfvt RightWall heat/flux/virial/tally LeftLiquid
    variable hflux_hfvt equal c_hflow_hfvt/(ly*lz)
 
    # x component of approximate heat flux vector inside the liquid region,
@@ -131,7 +131,7 @@
    variable press_ft equal c_rforce_ft/(ly*lz)
    #
    compute rforce_hfvt all reduce sum c_hflow_hfvt[1]
-   variable press_hfvt equal -c_rforce_hfvt/(ly*lz)
+   variable press_hfvt equal c_rforce_hfvt/(ly*lz)
 
 ----------    
```

Should I open a new PR or update this one?

akohlmey commented:
> Should I open a new PR or update this one?

The PR has already been merged, but you can push a commit to this branch and I can cherry-pick to some other pending PR like #2829

donatas-surblys commented:
I've committed the change to my heat-flux-virial-tally branch.

akohlmey commented:
> I've committed the change to my heat-flux-virial-tally branch.

Thanks. I've added it to the pending pull request.
